libbeat/common/kubernetes/util.go

```diff
@@ -94,54 +98,65 @@ func IsInCluster(kubeconfig string) bool {
 // If host is provided in the config use it directly.
 // If beat is deployed in k8s cluster, use hostname of pod which is pod name to query pod meta for node name.
 // If beat is deployed outside k8s cluster, use machine-id to match against k8s nodes for node name.
-func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (node string) {
+// If node cannot be discovered, return NODE_NAME env var as default value. In case it is not set return error.
+func DiscoverKubernetesNode(log *logp.Logger, host string, inCluster bool, client kubernetes.Interface) (string, error) {
 	if host != "" {
 		log.Infof("kubernetes: Using node %s provided in the config", host)
-		return host
+		return host, nil
+	}
+	nodeNameEnv := os.Getenv("NODE_NAME")
+	var envError, logerror error
+	if nodeNameEnv == "" {
+		envError = errors.New("kubernetes: NODE_NAME environment variable was not set")
 	}
 	ctx := context.TODO()
 	if inCluster {
 		ns, err := InClusterNamespace()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get namespace when beat is in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
-		podName, err := os.Hostname()
+		podName, err := osHostname()
 		if err != nil {
-			log.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Couldn't get hostname as beat pod name in cluster with error: %+v", err.Error())
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
 		}
 		log.Infof("kubernetes: Using pod name %s and namespace %s to discover kubernetes node", podName, ns)
 		pod, err := client.CoreV1().Pods(ns).Get(ctx, podName, metav1.GetOptions{})
 		if err != nil {
-			log.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
-			return defaultNode
+			logerror = fmt.Errorf("kubernetes: Querying for pod failed with error: %+v", err)
+			log.Error(logerror)
+			return nodeNameEnv, errors.Wrap(envError, logerror.Error())
```

Comment:
Does this work as expected? I would expect that if some of the methods tried fail, it logs something at the debug level and continue. Here it seems that when any of the methods fail, the whole discovery fails.
I would suggest to refactor this to a sequence of discovery methods tried one after the other. If one works, its result is returned, otherwise the error is logged at the debug level and the function continues.
For simplicity and readability, each one of the discovery methods could possibly go to its own function.

Reply:
Thanks a lot @jsoriano for taking a look. I have done a big refectory of the legacy `DiscoverKubernetesNode`. Regarding your comment I will sum up my opinion:

`DiscoverKubernetesNode` fails in specific cases:
Let's divide them in two categories:
1. `isInCluster` true meaning that kubeconfig exists and is valid. In this scenario `Node` will be retrieved from the pod code is running on. Failing cases are:

-  namespace cannot be found. In that case if namespace is not found we cannot move to next step which is to get the pod in that namespace and get the node it is running on. So it has to fail

-  podname cannot be retrieved from os.Hostname. Same as above without the pod name we cannot find the pod and the node it is running on
-  If namespace and podname exist but getting the actual pod from kubernetes fails. Without the pod it cannot get the node

2. `isInCluster` false. kubeconfig cannot be found or invalid. Code tries to get the `Node` needed via `machineId` of the host it running on. It fails when
a.  MachineId is "". Without a valid value it cannot find a kubernetes node with same MachineId. So it has to fail
b. It fails to list the available kubernetes nodes. Without the nodes it cannot compare the `MachineId` against the node's `MachineId`
c. No Node from the ones retrieved has tha MachineId. 

In all above scenarios discovery won't fail if `NODE_NAME` env var is set.

Reply:
Sorry, first of all, I was wrong about the behaviour, I thought that an error was returned in any case, but this wouldn't happen with `errors.Wrap()` when `envError` is nil. But, someone misreading this code is a bad signal in any case :slightly_smiling_face: There were three errors in the scope, plus the use of wrapping is not the best for readability.

After your refactor this is a bit simpler now, only two errors in the scope, but I still think that we have the opportunity to further improve it.

Some ideas below, please look at them as possible opportunities to improve the readability and maintainability of this code a little bit more, nothing blocker, this is already adding some nice things: error handling and a better fallback method for node discovery.

* You mentioned two clear categories, I think they can be two different functions, one called after the other, this would likely simplify the code.
* It can make sense to try the "machine id" method after the in-cluster one. In that case the use of the environment variable would only be the next thing to try.
* Imagine what you would need to do with current code if you wanted to add another fallback if the environment variable was also not available.
* I think that we should still log something at the debug level if discovery fails and we fallback to the use of `NODE_NAME`. In that case maybe the user has forgotten to explicitly set `host`, or they don't expect the use of `NODE_NAME`, and it may be difficult to investigate as with current code nothing would be logged.
* Not sure about the interface added, I wouldn't expect multiple implementations of this, and it doesn't seem to be leveraged in tests. It looks a bit artificial to me.